File: backend/app/schemas/dag.py
"""DAG schemas."""
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict


class DagNodeCreate(BaseModel):
    key: str
    name: str
    node_type: str  # input_value|calculation|distribution|validation
    stream: str = "distribution"
    formula: str | None = None
    description: str | None = None
    input_source: str | None = None
    variable_id: int | None = None
    payment_type: str | None = None
    export_field: str | None = None
    tolerance: Decimal | None = None
    tolerance_type: str | None = None
    comparison_variable: str | None = None
    default_prior_value: Decimal | None = None
    position_x: int = 0
    position_y: int = 0


class DagEdgeCreate(BaseModel):
    source_key: str
    target_key: str


class DagSaveRequest(BaseModel):
    nodes: list[DagNodeCreate]
    edges: list[DagEdgeCreate]
    description: str | None = None


class DagNodeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    key: str
    name: str
    node_type: str
    stream: str
    formula: str | None
    description: str | None
    input_source: str | None
    variable_id: int | None
    payment_type: str | None
    export_field: str | None
    tolerance: Decimal | None
    tolerance_type: str | None
    comparison_variable: str | None
    default_prior_value: Decimal | None
    position_x: int
    position_y: int
    is_active: bool


class DagEdgeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    source_node_id: int
    target_node_id: int


class DagVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    deal_id: int
    version_number: int
    description: str | None
    created_by: str
    created_at: datetime
    is_current: bool


class DagLoadResponse(BaseModel):
    version: DagVersionResponse
    nodes: list[DagNodeResponse]
    edges: list[DagEdgeResponse]

File: frontend/src/api/dag.ts
import { api } from "./client";

/* ── Types ─────────────────────────────────────────────────── */

export interface DagNode {
  id: number;
  key: string;
  name: string;
  node_type: string;
  stream: string;
  formula: string | null;
  description: string | null;
  input_source: string | null;
  variable_id: number | null;
  payment_type: string | null;
  export_field: string | null;
  tolerance: number | null;
  tolerance_type: string | null;
  comparison_variable: string | null;
  default_prior_value: number | null;
  waterfall_order: number | null;
  position_x: number;
  position_y: number;
  is_active: boolean;
}

export interface DagEdge {
  id: number;
  source_node_id: number;
  target_node_id: number;
}

export interface DagVersion {
  id: number;
  deal_id: number;
  version_number: number;
  description: string | null;
  created_by: string;
  created_at: string;
  is_current: boolean;
}

/* ── DAG load (returns current snapshot) ───────────────────── */

interface DagLoadResponse {
  version: DagVersion;
  nodes: DagNode[];
  edges: DagEdge[];
}

/* ── Queries ───────────────────────────────────────────────── */

export async function fetchNodes(dealId: number): Promise<DagNode[]> {
  const res = await api.get<DagLoadResponse>(`/deals/${dealId}/dag`);
  return res.nodes;
}

export async function fetchEdges(dealId: number): Promise<DagEdge[]> {
  const res = await api.get<DagLoadResponse>(`/deals/${dealId}/dag`);
  return res.edges;
}

export async function fetchVersions(dealId: number): Promise<DagVersion[]> {
  return api.get<DagVersion[]>(`/deals/${dealId}/dag/versions`);
}

/* ── Node mutations ────────────────────────────────────────── */

export function createNode(dealId: number, payload: Partial<DagNode>): Promise<DagNode> {
  return api.post<DagNode>(`/deals/${dealId}/dag/nodes`, payload);
}

export function updateNode(nodeId: number, fields: Record<string, unknown>, dealId?: number): Promise<DagNode> {
  if (dealId) {
    return api.patch<DagNode>(`/deals/${dealId}/dag/nodes/${nodeId}`, fields);
  }
  return api.patch<DagNode>(`/dag/nodes/${nodeId}`, fields);
}

export function deleteNode(nodeId: number): Promise<void> {
  return api.del(`/dag/nodes/${nodeId}`) as Promise<void>;
}

export function deactivateNode(nodeId: number): Promise<void> {
  return api.patch<void>(`/dag/nodes/${nodeId}/deactivate`, {});
}

export function reactivateNode(nodeId: number): Promise<void> {
  return api.patch<void>(`/dag/nodes/${nodeId}/reactivate`, {});
}

/* ── Edge mutations ────────────────────────────────────────── */

export function createEdge(
  dealId: number,
  payload: { source_node_id: number; target_node_id: number },
): Promise<DagEdge> {
  return api.post<DagEdge>(`/deals/${dealId}/dag/edges`, payload);
}

export function deleteEdge(edgeId: number): Promise<void> {
  return api.del(`/dag/edges/${edgeId}`) as Promise<void>;
}

/* ── Version mutations ─────────────────────────────────────── */

export function saveDag(
  dealId: number,
  description: string,
): Promise<{ version_id: number; version_number: number }> {
  return api.post(`/deals/${dealId}/dag`, { description });
}

export function revertDag(
  dealId: number,
  versionId: number,
): Promise<{ version_id: number; version_number: number }> {
  return api.post(`/deals/${dealId}/dag/revert/${versionId}`);
}

export function importDag(
  dealId: number,
  payload: Record<string, unknown>,
): Promise<{ version_id: number; version_number: number }> {
  return api.post(`/deals/${dealId}/dag/import`, payload);
}
